transports/webtransport-websys/src/connection.rs

```diff
@@ -0,0 +1,184 @@
+use libp2p_core::multihash::Multihash;
+use libp2p_core::muxing::{StreamMuxer, StreamMuxerEvent, StreamMuxerExt};
+use libp2p_core::{OutboundUpgrade, UpgradeInfo};
+use libp2p_identity::{Keypair, PeerId};
+use send_wrapper::SendWrapper;
+use std::collections::HashSet;
+use std::future::poll_fn;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{ReadableStreamDefaultReader, WebTransport, WebTransportBidirectionalStream};
+
+use crate::endpoint::Endpoint;
+use crate::fused_js_promise::FusedJsPromise;
+use crate::utils::{parse_reader_response, to_js_type};
+use crate::{Error, Stream};
+
+/// An opened WebTransport connection.
+pub struct Connection {
+    session: SendWrapper<WebTransport>,
+    create_stream_promise: FusedJsPromise,
+    incoming_stream_promise: FusedJsPromise,
+    incoming_streams_reader: SendWrapper<ReadableStreamDefaultReader>,
+    closed: bool,
+}
+
+impl Connection {
+    pub(crate) fn new(endpoint: &Endpoint) -> Result<Self, Error> {
+        let url = endpoint.url();
+
+        let session = if endpoint.certhashes.is_empty() {
+            // Endpoint has CA-signed TLS certificate
+            WebTransport::new(&url).map_err(Error::from_js_value)?
+        } else {
+            // Endpoint has self-signed TLS certificates
+            let opts = endpoint.webtransport_opts();
+            WebTransport::new_with_options(&url, &opts).map_err(Error::from_js_value)?
+        };
+
+        let incoming_streams = session.incoming_bidirectional_streams();
+        let incoming_streams_reader =
+            to_js_type::<ReadableStreamDefaultReader>(incoming_streams.get_reader())?;
+
+        Ok(Connection {
+            session: SendWrapper::new(session),
+            create_stream_promise: FusedJsPromise::new(),
+            incoming_stream_promise: FusedJsPromise::new(),
+            incoming_streams_reader: SendWrapper::new(incoming_streams_reader),
+            closed: false,
+        })
+    }
+
+    pub(crate) async fn authenticate(
+        &mut self,
+        keypair: &Keypair,
+        remote_peer: Option<PeerId>,
+        certhashes: HashSet<Multihash>,
+    ) -> Result<PeerId, Error> {
+        self.ready().await?;
+        let stream = self.create_stream().await?;
+
+        let mut noise = libp2p_noise::Config::new(keypair)?;
+
+        if !certhashes.is_empty() {
+            noise = noise.with_webtransport_certhashes(certhashes);
+        }
+
+        // We do not use `upgrade::apply_outbound` function because it uses
+        // `multistream_select` protocol, which is not used by WebTransport spec.
+        let info = noise.protocol_info().next().unwrap_or_default();
+        let (peer_id, _io) = noise.upgrade_outbound(stream, info).await?;
+
+        // TODO: This should be part libp2p-noise
+        if let Some(expected_peer_id) = remote_peer {
```

Comment:
Related: https://github.com/libp2p/rust-libp2p/issues/2946

To actually make use of this in our upgrade stack, some more changes are required.

Reply:
@thomaseizinger I was actually thinking to do the following:

* Add `libp2p_noise::Config::with_remote_peer_id`
* The above will pass the value at `libp2p_noise::io::handshake::State::expected_remote_peer_id`
* In `libp2p_noise::io::handshake::State::finish` we will verify if `id_pk.to_peer_id()` is the same with `State::expected_remote_peer_id`

After this change I would be able to do some like this in the `Connection::authenticate`:

```rust
        let mut noise = libp2p_noise::Config::new(keypair)?;

        if !certhashes.is_empty() {
            noise = noise.with_webtransport_certhashes(certhashes);
        }

        if let Some(expected_peer_id) = remote_peer {
            noise = noise.with_remote_peer_id(expected_peer_id);
        }
```

What do you think? I can do it in separate PR

Reply:
Yes sounds good to me!

I only linked https://github.com/libp2p/rust-libp2p/issues/2946 here because it would be great if we can take advantage of this built-in verification for all connection upgrades and not just here.

Reply:
> What do you think? I can do it in separate PR

Preference for separate pull request.